### PR TITLE
Feature/list generator

### PR DESCRIPTION
**Pull request by faneshion**
#310

### CODE REVIEW

File: matchzoo/generators/list_generator.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    def transform_relation(self, relations: pd.DataFrame) -> list:
+        """Obtain the transformed data from :class:`DataPack`.
+
+        Note here, label is required to make lists.
+
+        :param relations: An instance of DataFrame to be transformed.
+        :return: the output of all the lists' indices.
+        """
+        # Note here the main id is set to be the id_left
+        lists = []
```
Comment (bwanglzu) [inline matchzoo/generators/list_generator.py]: change the variable name.

File: matchzoo/generators/list_generator.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        :param batch_size: number of instances in a batch.
+        :param stage: the current phase, the value can be 'train' or 'test'.
+        :param shuffle: whether to shuffle the instances while generating a
+            batch.
+        """
+        self._left = inputs.left
+        self._right = inputs.right
+        self._relation = inputs.relation
+        self._task = tasks.Ranking()
+        self._lists = self.transform_relation(self._relation)
```
Comment (bwanglzu) [inline matchzoo/generators/list_generator.py]: change the variable name

File: matchzoo/generators/list_generator.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    def _get_batch_of_transformed_samples(
+        self,
+        index_array: np.array
+    ) -> typing.Tuple[dict, typing.Any]:
+        """Get a batch of samples based on their ids.
+
+        :param index_array: a list of instance ids.
+        :return: A batch of transformed samples.
+        """
+        trans_index = self._lists[index_array[0]]
```
Comment (bwanglzu) [inline matchzoo/generators/list_generator.py]: not clear what you're doing here

File: matchzoo/generators/list_generator.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        inputs: datapack.DataPack,
+        batch_size: int=1,
+        stage: str='train',
+        shuffle: bool=True
+    ):
+        """Construct the list generator.
+
+        :param inputs: the output generated by :class:`DataPack`.
+        :param batch_size: number of instances in a batch.
+        :param stage: the current phase, the value can be 'train' or 'test'.
```
Comment (bwanglzu) [inline matchzoo/generators/list_generator.py]: 'train' -> \`train\`

File: matchzoo/generators/list_generator.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        batch_y = None
+        if self.stage == 'train':
+            self._relation['label'] = self._relation['label'].astype(
+                self._task.output_dtype)
+            batch_y = self._relation.iloc[trans_index, 2].values
+
+        columns = self._left.columns.values.tolist() + \
+            self._right.columns.values.tolist() + ['ids']
+        for column in columns:
+            batch_x[column] = []
```
Comment (bwanglzu) [inline matchzoo/generators/list_generator.py]: use a list compression

File: matchzoo/generators/list_generator.py (hunk 4 above)
Comment (faneshion) [reply matchzoo/generators/list_generator.py]: done~

File: matchzoo/generators/list_generator.py (hunk 5 above)
Comment (faneshion) [reply matchzoo/generators/list_generator.py]: done~

File: matchzoo/generators/list_generator.py (hunk 3 above)
Comment (faneshion) [reply matchzoo/generators/list_generator.py]: Right now in this version~ The batch_size is set to be 1. Thus, the `index_array` is a list which contains only one element.

File: matchzoo/generators/list_generator.py (hunk 1 above)
Comment (faneshion) [reply matchzoo/generators/list_generator.py]: Change to `id_lists`~

File: tests/unit_test/generators/test_list_generator.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    assert x0['text_left'].tolist() == [[1, 2]]
+    assert x0['text_right'].tolist() == [[2, 3, 4]]
+    assert x1['ids'].tolist() == [['qid1', 'did1'], ['qid1', 'did0']]
+    assert x1['text_left'].tolist() == [[2, 3], [2, 3]]
+    assert x1['text_right'].tolist() == [[3, 4, 5], [2, 3, 4]]
+    if stage == 'test':
+        assert y0 is None
+        assert y1 is None
+    elif stage == 'train':
+        assert y0.tolist() == [0.0]
```
[the remark this answers is not in the document]
Comment (faneshion) [reply tests/unit_test/generators/test_list_generator.py]: task `Ranking` has defined the `output_dtype` as `float`~

**Pull request merged**